src/transformers/integrations.py

```diff
@@ -926,74 +932,257 @@ def __del__(self):
+            [neptune.init_run()](https://docs.neptune.ai/api-reference/neptune#.init_run) function when a new run is
+            created.
     """
 
-    def __init__(self):
+    if TYPE_CHECKING and is_neptune_available():
+        from neptune.new.metadata_containers.run import Run
+
+    class MissingConfiguration(Exception):
+        def __init__(self):
+            super().__init__(
+                """
+            ------ Unsupported ---- We were not able to create new runs. You provided a custom Neptune run to
+            `NeptuneCallback` with the `run` argument. For the integration to work fully, provide your `api_token` and
+            `project` by saving them as environment variables or passing them to the callback.
+            """
+            )
+
+    INTEGRATION_VERSION_KEY = "source_code/integrations/transformers"
+    MODEL_PARAMETERS_KEY = "model_parameters"
+    TRIAL_NAME_KEY = "trial"
+    TRIAL_PARAMS_KEY = "trial_params"
+    TRAINER_PARAMETERS_KEY = "trainer_parameters"
+    FLAT_METRICS = {"train/epoch"}
+
+    def __init__(
+        self,
+        *,
+        api_token: Optional[str] = None,
+        project: Optional[str] = None,
+        name: Optional[str] = None,
+        base_namespace: str = "finetuning",
+        run: Optional["Run"] = None,
+        log_parameters: bool = True,
+        log_checkpoints: Optional[str] = None,
+        **neptune_run_kwargs
+    ):
         if not is_neptune_available():
             raise ValueError(
-                "NeptuneCallback requires neptune-client to be installed. Run `pip install neptune-client`."
+                "NeptuneCallback requires the Neptune client library to be installed. "
+                "To install the library, run `pip install neptune-client`."
             )
-        import neptune.new as neptune
 
-        self._neptune = neptune
-        self._initialized = False
-        self._log_artifacts = False
+        from neptune.new.metadata_containers.run import Run
 
-    def setup(self, args, state, model):
-        """
-        Setup the Neptune integration.
+        try:
+            from neptune.new.integrations.utils import verify_type
+        except ImportError:
+            from neptune.new.internal.utils import verify_type
+
+        verify_type("api_token", api_token, (str, type(None)))
+        verify_type("project", project, (str, type(None)))
+        verify_type("name", name, (str, type(None)))
+        verify_type("base_namespace", base_namespace, str)
+        verify_type("run", run, (Run, type(None)))
+        verify_type("log_parameters", log_parameters, bool)
+        verify_type("log_checkpoints", log_checkpoints, (str, type(None)))
+
+        self._base_namespace_path = base_namespace
+        self._log_parameters = log_parameters
+        self._log_checkpoints = log_checkpoints
+        self._initial_run: Optional[Run] = run
+
+        self._run: Optional[Run] = None
```

Comment:
We don't use type hints except when it really explains what an obscure variable is. Here the naming makes it perfectly clear it's a `Run`.

Reply:
Sure. Done :wink: